llm/cpp/group_beam_searcher.hpp

```diff
@@ -0,0 +1,239 @@
+    std::function<bool(const Beam&)> early_finish = [](const Beam&){return false;};
+};
+
+struct Group {
+    std::vector<Beam> ongoing;  // Best beams in front
+    std::vector<Beam> min_heap;  // The worst of the best completed beams is the first
+    bool done = false;
+    void finish(Beam&& beam, const Parameters& parameters) {
+        beam.score /= std::pow(float(parameters.prompt.size() + beam.tokens.size()), parameters.length_penalty);
+        min_heap.push_back(std::move(beam));
+        std::push_heap(min_heap.begin(), min_heap.end(), greater);
+        if (min_heap.size() > parameters.group_size) {
+            std::pop_heap(min_heap.begin(), min_heap.end(), greater);
+            min_heap.pop_back();
+        }
+    }
+    void is_done(const Parameters& parameters) {
+        if (min_heap.size() < parameters.group_size) {
+            return;
+        }
+        size_t cur_len = parameters.prompt.size() + ongoing.front().tokens.size();
+        float best_sum_logprobs = ongoing.front().score;
+        float worst_score = min_heap.front().score;
+        switch (parameters.stop_criteria) {
+            case StopCriteria::early:
+                done = true;
+                return;
+            case StopCriteria::heuristic: {
+                float highest_attainable_score = best_sum_logprobs / std::pow(float(cur_len), parameters.length_penalty);
+                done = worst_score >= highest_attainable_score;
+                return;
+            }
+            case StopCriteria::never: {
+                size_t length = parameters.length_penalty > 0.0 ? parameters.max_new_tokens : cur_len;
+                float highest_attainable_score = best_sum_logprobs / std::pow(float(length), parameters.length_penalty);
+                done = worst_score >= highest_attainable_score;
+                return;
+            }
+            default: throw std::runtime_error("Never reached");
+        }
+    }
+};
+
+struct TokenToBeam {int64_t token_idx; size_t beam_idx;};
+
+struct GroupBeamSearcher {
+    Parameters parameters;
+    std::vector<Group> groups;
+    GroupBeamSearcher(Parameters parameters) : parameters{std::move(parameters)}, groups{parameters.n_groups} {
+        if (parameters.no_repeat_ngram_size == 0) {
+            throw std::runtime_error("no_repeat_ngram_size must be positive");
+        }
+        for (Group& group : groups) {
+            group.ongoing.resize(parameters.group_size);
+            group.ongoing.front().score = 0.0;
+        }
+    }
+    std::vector<TokenToBeam> process(const ov::Tensor& logits) {
+        std::vector<TokenToBeam> next_tokens;
+        next_tokens.reserve(parameters.n_groups * parameters.group_size);
+        size_t beam_count = 0;
+        for (Group& group : groups) {
+            if (!group.done) {
+                for (Beam& beam : group.ongoing) {
+                    beam.global_beam_idx = beam_count;
+                    if (!beam.tokens.empty() && logits.get_shape().at(0) != 1) {
```

Comment:
do we need condition `logits.get_shape().at(0) != 1`? 
looks like on the second iteration `batch_size = logits.get_shape().at(0)` is always non equal to 1 and at the same time beam.tokens contain a single token.